server/storage/wal/wal.go

```diff
@@ -396,6 +396,65 @@ func openAtIndex(lg *zap.Logger, dirpath string, snap walpb.Snapshot, write bool
 	return w, nil
 }
 
+func ReadAllEntries(lg *zap.Logger, dirpath string) (state raftpb.HardState, ents []raftpb.Entry, err error) {
+	names, nameIndex, err := selectWALFiles(lg, dirpath, walpb.Snapshot{Term: 0, Index: 0})
+	if err != nil {
+		return state, nil, fmt.Errorf("[ReadAllEntries] selectWALFiles failed: %w", err)
+	}
+
+	rs, _, closer, err := openWALFiles(lg, dirpath, names, nameIndex, false)
+	if err != nil {
+		return state, nil, fmt.Errorf("[ReadAllEntries] openWALFiles failed: %w", err)
+	}
+	defer closer()
+	rec := &walpb.Record{}
+	decoder := NewDecoder(rs...)
+	var startIndex uint64
+	for err = decoder.Decode(rec); err == nil; err = decoder.Decode(rec) {
+		switch rec.Type {
+		case EntryType:
+			e := MustUnmarshalEntry(rec.Data)
+			if e.Index == 0 {
+				return state, nil, fmt.Errorf("entry with index 0 found")
+			}
+			if startIndex == 0 {
+				startIndex = e.Index
+				ents = append(ents, e)
+				continue
+			}
+			if e.Index > startIndex {
+				offset := e.Index - startIndex
+				if offset > uint64(len(ents)) {
+					ents = append(ents, make([]raftpb.Entry, offset-uint64(len(ents)))...)
+				}
+				// The line below is potentially overriding some 'uncommitted' entries.
+				ents = append(ents[:offset], e)
+			}
+		case StateType:
+			state = MustUnmarshalState(rec.Data)
+		case MetadataType:
+		case CrcType:
+		case SnapshotType:
+		default:
+			return state, nil, fmt.Errorf("unexpected block type %d", rec.Type)
+		}
+	}
+	if err != nil && !errors.Is(err, io.EOF) {
+		return state, ents, err
+	}
+	// Filter empty entries
+	i := 0
+	for _, e := range ents {
+		if e.Index == 0 {
+			continue
+		}
```

Comment:
Maybe it's just me, but the first I saw this line, I thought the code could not reach here because of the logic on line 417.  Then looking again, the empty entries are from offset handling on line 428.  

Perhaps we should make it clear on that line that line 428 pads the empty slots.  Not sure if something like below would make it clearer...:
```go
ents = append(ents, 
  slices.Repeat(raftpb.Entry{}, offset-uint64(len(ents))...)
```

Reply:
I don't think there is significant difference between `slices.Repeat(raftpb.Entry{}, offset-uint64(len(ents))` and `make([]raftpb.Entry, offset-uint64(len(ents)))` in making them more understandable. Allocating continuous slice and overriding uncommitted entries is the hard/magical part about the reading WAL. 

I think we could make it more clearer and even try to unify it with ReadAll, however that's not in scope of the PR.